### PR TITLE
RavenDB-18232 - Sharding - Edit Replication Sink View Endpoints

### PR DESCRIPTION
**Pull request by shiranshalom**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18232

### Type of change

- New feature

### How risky is the change?

- Not relevant

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- It has been verified by manual testing

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Documents/Handlers/Processors/Replication/AbstractPullReplicationHandlerProcessorForUpdatePullReplicationOnSinkNode.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+            responseJson[nameof(ModifyOngoingTaskResult.TaskId)] = _pullReplication.TaskId == 0 ? index : _pullReplication.TaskId;
+        }
+
+        protected override void OnBeforeUpdateConfiguration(ref BlittableJsonReaderObject configuration, JsonOperationContext context)
+        {
+            if (ResourceNameValidator.IsValidResourceName(GetDatabaseName(), RequestHandler.ServerStore.Configuration.Core.DataDirectory.FullPath, out string errorMessage) == false)
+                throw new BadRequestException(errorMessage);
+
+            RequestHandler.ServerStore.LicenseManager.AssertCanAddPullReplicationAsSink();
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Processors/Replication/AbstractPullReplicationHandlerProcessorForUpdatePullReplicationOnSinkNode.cs:46]: it should go to assert method

**Review by ml054 — APPROVED** [review]

File: src/Raven.Server/Web/System/Processors/OngoingTasks/OngoingTasksHandlerProcessorForGetOngoingTasks.cs [hunk 2]
```diff
@@ -51,7 +54,7 @@ protected override IEnumerable<OngoingTaskBackup> CollectBackupTasks(Transaction
 
         foreach (var backupConfiguration in databaseRecord.PeriodicBackups)
         {
-            yield return GetOngoingTaskBackup(backupConfiguration.TaskId, databaseRecord, backupConfiguration, clusterTopology);
+            yield return GetOngoingTaskBackupAsync(backupConfiguration.TaskId, databaseRecord, backupConfiguration, clusterTopology).Result;
```
Comment (karmeli87) [inline src/Raven.Server/Web/System/Processors/OngoingTasks/OngoingTasksHandlerProcessorForGetOngoingTasks.cs]: let's make this `IAsyncEnumerable<OngoingTaskBackup>` and use `await`

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/OngoingTasks/ShardedOngoingTasksHandlerProcessorForGetOngoingTasks.cs [hunk 3]
```diff
@@ -162,4 +175,100 @@ protected override int SubscriptionsCount
...
+            new GetOngoingTaskInfoOperation.GetOngoingTaskInfoCommand(key, type) : 
+            new GetOngoingTaskInfoOperation.GetOngoingTaskInfoCommand(taskName, type);
+
+        return new ProxyCommand<OngoingTask>(command, RequestHandler.HttpContext.Response);
+    }
+
+    protected override async ValueTask<OngoingTaskSubscription> GetSubscriptionTaskInfoAsync(DatabaseRecord record, ClusterTopology clusterTopology,
+        SubscriptionState subscriptionState, long key)
+    {
+        DevelopmentHelper.ShardingToDo(DevelopmentHelper.TeamMember.Shiran, DevelopmentHelper.Severity.Normal, "implement for sharding - https://issues.hibernatingrhinos.com/issue/RavenDB-13113");
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Sharding/Handlers/Processors/OngoingTasks/ShardedOngoingTasksHandlerProcessorForGetOngoingTasks.cs]: what does it mean?

Comment (karmeli87) [reply src/Raven.Server/Documents/Sharding/Handlers/Processors/OngoingTasks/ShardedOngoingTasksHandlerProcessorForGetOngoingTasks.cs]: In general I think it is fine to pass nodeTag&shardNumber to be used as a proxy..
But for the sharded handler I think it should be okay also to not pass those parameters and get only the definition from the record directly..

File: src/Raven.Server/Web/System/Processors/OngoingTasks/AbstractOngoingTasksHandlerProcessorForGetOngoingTasks.cs [hunk 4]
```diff
@@ -22,6 +38,12 @@ protected AbstractOngoingTasksHandlerProcessorForGetOngoingTasks([NotNull] TRequ
 
     public override async ValueTask ExecuteAsync()
     {
+        if (RequestHandler.HttpContext.Request.Path == $"/databases/{RequestHandler.DatabaseName}/task")
```
Comment (karmeli87) [inline src/Raven.Server/Web/System/Processors/OngoingTasks/AbstractOngoingTasksHandlerProcessorForGetOngoingTasks.cs]: what is that?

File: src/Raven.Server/Web/System/Processors/OngoingTasks/OngoingTasksHandlerProcessorForGetOngoingTasks.cs [hunk 5]
```diff
@@ -297,11 +260,62 @@ protected override IEnumerable<OngoingTaskReplication> CollectExternalReplicatio
...
+                TaskState = subscriptionState.Disabled ? OngoingTaskState.Disabled : OngoingTaskState.Enabled,
+                MentorNode = subscriptionState.MentorNode,
+                TaskId = subscriptionState.SubscriptionId,
+                Query = subscriptionState.Query,
+                TaskConnectionStatus = connectionStatus
+            };
+        }
+    }
+
+    protected override ValueTask<OngoingTaskReplication> GetExternalReplicationInfoAsync(DatabaseTopology databaseTopology, ClusterTopology clusterTopology,
```
Comment (karmeli87) [inline src/Raven.Server/Web/System/Processors/OngoingTasks/OngoingTasksHandlerProcessorForGetOngoingTasks.cs]: We need to split those methods into 2 part.
part 1: get definition -> same code for sharded/non-sharded
part 2: get status -> this is a "local" info which we will implement only for the non-sharded version

**Review by karmeli87 — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/OngoingTasks/ShardedOngoingTasksHandlerProcessorForGetOngoingTasks.cs (hunk 3 above)
Comment (ppekrol) [reply src/Raven.Server/Documents/Sharding/Handlers/Processors/OngoingTasks/ShardedOngoingTasksHandlerProcessorForGetOngoingTasks.cs]: Ping me about this

File: src/Raven.Server/Web/System/Processors/OngoingTasks/AbstractOngoingTasksHandlerProcessorForGetOngoingTasks.cs (hunk 4 above)
Comment (shiranshalom) [reply src/Raven.Server/Web/System/Processors/OngoingTasks/AbstractOngoingTasksHandlerProcessorForGetOngoingTasks.cs]: done

File: src/Raven.Server/Web/System/Processors/OngoingTasks/OngoingTasksHandlerProcessorForGetOngoingTasks.cs (hunk 5 above)
Comment (shiranshalom) [reply src/Raven.Server/Web/System/Processors/OngoingTasks/OngoingTasksHandlerProcessorForGetOngoingTasks.cs]: done

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/OngoingTasks/ShardedOngoingTasksHandlerProcessorForGetOngoingTasks.cs [hunk 6]
```diff
@@ -1,23 +1,58 @@
...
 using Raven.Server.ServerWide;
 using Raven.Server.ServerWide.Context;
 using Raven.Server.Web.System.Processors.OngoingTasks;
+using Sparrow.Json;
 using Sparrow.Utils;
 
 namespace Raven.Server.Documents.Sharding.Handlers.Processors.OngoingTasks;
 
-internal class ShardedOngoingTasksHandlerProcessorForGetOngoingTasks : AbstractOngoingTasksHandlerProcessorForGetOngoingTasks<ShardedDatabaseRequestHandler, TransactionOperationContext>
+internal class ShardedOngoingTasksHandlerProcessorForGetOngoingTaskInfo : ShardedOngoingTasksHandlerProcessorForGetOngoingTasksInfo
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Handlers/Processors/OngoingTasks/ShardedOngoingTasksHandlerProcessorForGetOngoingTasks.cs]: move those to separate files

File: src/Raven.Server/Web/System/Processors/OngoingTasks/AbstractOngoingTasksHandlerProcessorForGetOngoingTasks.cs [hunk 7]
```diff
@@ -1,37 +1,45 @@
...
-
-        using (ContextPool.AllocateOperationContext(out JsonOperationContext context))
-        await using (var writer = new AsyncBlittableJsonTextWriter(context, RequestHandler.ResponseBodyStream()))
-        {
-            context.Write(writer, result.ToJson());
-        }
-    }
-
-    public OngoingTasksResult GetOngoingTasksInternal()
+    protected OngoingTasksResult GetOngoingTasksInternal()
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/Processors/OngoingTasks/AbstractOngoingTasksHandlerProcessorForGetOngoingTasks.cs]: Async

File: src/Raven.Server/Web/System/Processors/OngoingTasks/AbstractOngoingTasksHandlerProcessorForGetOngoingTasks.cs [hunk 8]
```diff
@@ -73,6 +81,207 @@ public OngoingTasksResult GetOngoingTasksInternal()
         }
     }
 
+    protected async ValueTask GetOngoingTaskInfoInternal()
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/Processors/OngoingTasks/AbstractOngoingTasksHandlerProcessorForGetOngoingTasks.cs]: Async

File: src/Raven.Server/Web/System/Processors/OngoingTasks/AbstractOngoingTasksHandlerProcessorForGetOngoingTasks.cs [hunk 9]
```diff
@@ -73,6 +81,207 @@ public OngoingTasksResult GetOngoingTasksInternal()
...
+    {
+        if (ResourceNameValidator.IsValidResourceName(RequestHandler.DatabaseName, ServerStore.Configuration.Core.DataDirectory.FullPath, out string errorMessage) == false)
+        {
+            bool sharded = ShardHelper.IsShardedName(RequestHandler.DatabaseName);
+            if (sharded == false)
+                throw new BadRequestException(errorMessage);
+        }
+    }
+
+    private async Task WriteResult(JsonOperationContext context, IDynamicJson taskInfo)
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/Processors/OngoingTasks/AbstractOngoingTasksHandlerProcessorForGetOngoingTasks.cs]: Async

File: src/Raven.Server/Web/System/Processors/OngoingTasks/OngoingTasksHandlerProcessorForGetOngoingTasks.cs [hunk 10]
```diff
@@ -23,16 +21,48 @@
 using Raven.Server.Documents.Replication.Outgoing;
 using Raven.Server.ServerWide;
 using Raven.Server.ServerWide.Context;
+using Sparrow.Json;
 
 namespace Raven.Server.Web.System.Processors.OngoingTasks;
 
-internal class OngoingTasksHandlerProcessorForGetOngoingTasks : AbstractOngoingTasksHandlerProcessorForGetOngoingTasks<DatabaseRequestHandler, DocumentsOperationContext>
+
+internal class OngoingTasksHandlerProcessorForGetOngoingTaskInfo : OngoingTasksHandlerProcessorForGetOngoingTasksInfo
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/Processors/OngoingTasks/OngoingTasksHandlerProcessorForGetOngoingTasks.cs]: move to separate files

**Comment by ppekrol** [comment]
It does not compile

File: src/Raven.Server/Web/System/Processors/OngoingTasks/AbstractOngoingTasksHandlerProcessorForGetOngoingTasks.cs (hunk 8 above)
Comment (shiranshalom) [reply src/Raven.Server/Web/System/Processors/OngoingTasks/AbstractOngoingTasksHandlerProcessorForGetOngoingTasks.cs]: done

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/OngoingTasks/ShardedOngoingTasksHandlerProcessorForGetOngoingTasks.cs (hunk 6 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Handlers/Processors/OngoingTasks/ShardedOngoingTasksHandlerProcessorForGetOngoingTasks.cs]: done

File: src/Raven.Server/Web/System/Processors/OngoingTasks/OngoingTasksHandlerProcessorForGetOngoingTasks.cs (hunk 10 above)
Comment (shiranshalom) [reply src/Raven.Server/Web/System/Processors/OngoingTasks/OngoingTasksHandlerProcessorForGetOngoingTasks.cs]: done

File: src/Raven.Server/Web/System/Processors/OngoingTasks/AbstractOngoingTasksHandlerProcessorForGetOngoingTasks.cs (hunk 9 above)
Comment (shiranshalom) [reply src/Raven.Server/Web/System/Processors/OngoingTasks/AbstractOngoingTasksHandlerProcessorForGetOngoingTasks.cs]: done

**Pull request merged**